controllers/pod.go

```diff
@@ -185,155 +160,227 @@ func (r *SingleClusterReconciler) quickRestart(
+
+		r.Log.V(1).Info("Pod deleted", "podName", pod.Name)
+	}
+	if len(restartedPods) > 0 {
+		return r.ensurePodsRunningAndReady(restartedPods)
+	}
+	return reconcileSuccess()
+}
+
+func (r *SingleClusterReconciler) ensurePodsRunningAndReady(podsToCheck []*corev1.Pod) reconcileResult {
+	podNames := getPodNames(podsToCheck)
+	readyPods := map[string]bool{}
+
+	const maxRetries = 6
+	const retryInterval = time.Second * 10
+
+	for i := 0; i < maxRetries; i++ {
+		r.Log.V(1).Info("Waiting for pods to be ready after delete", "pods", podNames)
+
+		for _, pod := range podsToCheck {
+			if readyPods[pod.Name] {
+				continue
+			}
 
-		if !utils.IsPodRunningAndReady(updatedPod) {
 			r.Log.V(1).Info(
-				"Waiting for pod to be ready", "podName", updatedPod.Name,
-				"status", updatedPod.Status.Phase, "DeletionTimestamp",
-				updatedPod.DeletionTimestamp,
+				"Waiting for pod to be ready", "podName", pod.Name,
+				"status", pod.Status.Phase, "DeletionTimestamp",
+				pod.DeletionTimestamp,
+			)
+
+			updatedPod := &corev1.Pod{}
+			podName := types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}
+			if err := r.Client.Get(context.TODO(), podName, updatedPod); err != nil {
+				return reconcileError(err)
+			}
+
+			if err := utils.CheckPodFailed(updatedPod); err != nil {
+				return reconcileError(err)
+			}
+
+			if !utils.IsPodRunningAndReady(updatedPod) {
+				break
+			}
+
+			readyPods[pod.Name] = true
+
+			r.Log.Info("Pod is restarted", "podName", updatedPod.Name)
+			r.Recorder.Eventf(r.aeroCluster, corev1.EventTypeNormal, "PodRestarted",
+				"[rack-%s] Restarted Pod %s", pod.Labels[asdbv1beta1.AerospikeRackIdLabel], pod.Name)
+		}
+
+		if len(readyPods) == len(podsToCheck) {
+			r.Log.Info(
+				"Pods are running and ready", "pods",
+				podNames,
 			)
-			time.Sleep(time.Second * 5)
+			return reconcileSuccess()
+		}
+		time.Sleep(retryInterval)
+	}
+
+	r.Log.Info(
+		"Timed out waiting for pods to come up", "pods",
+		podNames,
+	)
+	return reconcileRequeueAfter(10)
+}
+
+func getRearrangedFailedAndActivePods(pods []*corev1.Pod) []*corev1.Pod {
+	var rearrangedPods []*corev1.Pod
+
+	failedPods, activePods := getFailedAndActivePods(pods)
+	rearrangedPods = append(rearrangedPods, failedPods...)
+	rearrangedPods = append(rearrangedPods, activePods...)
+
+	return rearrangedPods
+}
+
+func getFailedAndActivePods(pods []*corev1.Pod) (failedPods, activePods []*corev1.Pod) {
+	for i := range pods {
+		pod := pods[i]
+
+		if err := utils.CheckPodFailed(pod); err != nil {
+			failedPods = append(failedPods, pod)
 			continue
 		}
+		activePods = append(activePods, pod)
+	}
+	return failedPods, activePods
+}
+
+func (r *SingleClusterReconciler) safelyDeletePodsAndEnsureImageUpdated(
+	rackState RackState, podsToUpdate []*corev1.Pod, ignorablePods []corev1.Pod,
+) reconcileResult {
 
-		r.Log.Info("Pod is restarted", "podName", updatedPod.Name)
-		started = true
-		break
+	failedPods, activePods := getFailedAndActivePods(podsToUpdate)
```

Comment:
can we design in such a way that getFailedAndActivePods will not be called again as we already using it in rearranging failed and active pods before this step.